Add support for bulk record updates, create domain fix

mintopia opened the pull request:
Based on the work in PR #18 and PR #19, this makes the following changes:

1. Add newly created domains to the cache so records can be created immediately.
2. Use bulk record creation/deletion endpoints on the DNS Made Easy API to reduce the number of API calls that need to be made to update records.

I have updated the tests to cover this and there is 100% test coverage.

yzguy commented:
@mintopia Hello again,

The zone is created fine and the records added immediately so that feature is good! 

In my testing, it seems like the bulk delete actually ends up deleting all the records not just the ones that should be deleted. Just some output to demonstrate this

```
-> octodns-sync --config-file=config.yaml --doit
2023-06-26T10:51:08  [4338451840] INFO  Manager __init__: config_file=config.yaml (octoDNS 0.9.21)
2023-06-26T10:51:08  [4338451840] INFO  Manager _config_executor: max_workers=1
2023-06-26T10:51:08  [4338451840] INFO  Manager _config_include_meta: include_meta=False
2023-06-26T10:51:08  [4338451840] INFO  Manager __init__: global_processors=[]
2023-06-26T10:51:08  [4338451840] INFO  Manager __init__: provider=config (octodns.provider.yaml 0.9.21)
2023-06-26T10:51:08  [4338451840] INFO  Manager __init__: provider=dnsmadeeasy (octodns_dnsmadeeasy 0.0.2+28d8e8b4)
2023-06-26T10:51:08  [4338451840] INFO  Manager sync: eligible_zones=[], eligible_targets=[], dry_run=False, force=False, plan_output_fh=<stdout>
2023-06-26T10:51:08  [4338451840] INFO  Manager sync:   zone=yzguy.lll.
2023-06-26T10:51:08  [4338451840] INFO  Manager sync:   sources=['config'] -> targets=['dnsmadeeasy']
2023-06-26T10:51:08  [4338451840] INFO  YamlProvider[config] populate:   found 3 records, exists=False
2023-06-26T10:51:08  [4338451840] INFO  DnsMadeEasyProvider[dnsmadeeasy] plan: desired=yzguy.lll.
2023-06-26T10:51:09  [4338451840] INFO  DnsMadeEasyProvider[dnsmadeeasy] populate:   found 0 records, exists=True
2023-06-26T10:51:09  [4338451840] WARNING DnsMadeEasyProvider[dnsmadeeasy] root NS record supported, but no record is configured for yzguy.lll.
2023-06-26T10:51:09  [4338451840] INFO  DnsMadeEasyProvider[dnsmadeeasy] plan:   Creates=3, Updates=0, Deletes=0, Existing Records=0
2023-06-26T10:51:09  [4338451840] INFO  Plan
********************************************************************************
* yzguy.lll.
********************************************************************************
* dnsmadeeasy (DnsMadeEasyProvider)
*   Create <ARecord A 3600, record1.yzguy.lll., ['192.168.225.10']> (config)
*   Create <ARecord A 3600, record2.yzguy.lll., ['192.168.225.11']> (config)
*   Create <ARecord A 3600, record3.yzguy.lll., ['192.168.225.12']> (config)
*   Summary: Creates=3, Updates=0, Deletes=0, Existing Records=0
********************************************************************************
```

Records get created fine in the UI

If I go and remove `record1`, the output shows only that would be removed

```
-> octodns-sync --config-file=config.yaml
2023-06-26T10:51:46  [4377822592] INFO  Manager __init__: config_file=config.yaml (octoDNS 0.9.21)
2023-06-26T10:51:46  [4377822592] INFO  Manager _config_executor: max_workers=1
2023-06-26T10:51:46  [4377822592] INFO  Manager _config_include_meta: include_meta=False
2023-06-26T10:51:46  [4377822592] INFO  Manager __init__: global_processors=[]
2023-06-26T10:51:46  [4377822592] INFO  Manager __init__: provider=config (octodns.provider.yaml 0.9.21)
2023-06-26T10:51:46  [4377822592] INFO  Manager __init__: provider=dnsmadeeasy (octodns_dnsmadeeasy 0.0.2+28d8e8b4)
2023-06-26T10:51:46  [4377822592] INFO  Manager sync: eligible_zones=[], eligible_targets=[], dry_run=True, force=False, plan_output_fh=<stdout>
2023-06-26T10:51:46  [4377822592] INFO  Manager sync:   zone=yzguy.lll.
2023-06-26T10:51:46  [4377822592] INFO  Manager sync:   sources=['config'] -> targets=['dnsmadeeasy']
2023-06-26T10:51:46  [4377822592] INFO  YamlProvider[config] populate:   found 2 records, exists=False
2023-06-26T10:51:46  [4377822592] INFO  DnsMadeEasyProvider[dnsmadeeasy] plan: desired=yzguy.lll.
2023-06-26T10:51:47  [4377822592] INFO  DnsMadeEasyProvider[dnsmadeeasy] populate:   found 3 records, exists=True
2023-06-26T10:51:47  [4377822592] WARNING DnsMadeEasyProvider[dnsmadeeasy] root NS record supported, but no record is configured for yzguy.lll.
2023-06-26T10:51:47  [4377822592] INFO  DnsMadeEasyProvider[dnsmadeeasy] plan:   Creates=0, Updates=0, Deletes=1, Existing Records=3
2023-06-26T10:51:47  [4377822592] INFO  Plan
********************************************************************************
* yzguy.lll.
********************************************************************************
* dnsmadeeasy (DnsMadeEasyProvider)
*   Delete <ARecord A 3600, record1.yzguy.lll., ['192.168.225.10']>
*   Summary: Creates=0, Updates=0, Deletes=1, Existing Records=3
********************************************************************************
```

However, when I go into the UI there are none of the records. Which is also shown if I run the noop again

```
-> octodns-sync --config-file=config.yaml
2023-06-26T10:54:21  [4315792768] INFO  Manager __init__: config_file=config.yaml (octoDNS 0.9.21)
2023-06-26T10:54:21  [4315792768] INFO  Manager _config_executor: max_workers=1
2023-06-26T10:54:21  [4315792768] INFO  Manager _config_include_meta: include_meta=False
2023-06-26T10:54:21  [4315792768] INFO  Manager __init__: global_processors=[]
2023-06-26T10:54:21  [4315792768] INFO  Manager __init__: provider=config (octodns.provider.yaml 0.9.21)
2023-06-26T10:54:21  [4315792768] INFO  Manager __init__: provider=dnsmadeeasy (octodns_dnsmadeeasy 0.0.2+28d8e8b4)
2023-06-26T10:54:21  [4315792768] INFO  Manager sync: eligible_zones=[], eligible_targets=[], dry_run=True, force=False, plan_output_fh=<stdout>
2023-06-26T10:54:21  [4315792768] INFO  Manager sync:   zone=yzguy.lll.
2023-06-26T10:54:21  [4315792768] INFO  Manager sync:   sources=['config'] -> targets=['dnsmadeeasy']
2023-06-26T10:54:21  [4315792768] INFO  YamlProvider[config] populate:   found 2 records, exists=False
2023-06-26T10:54:21  [4315792768] INFO  DnsMadeEasyProvider[dnsmadeeasy] plan: desired=yzguy.lll.
2023-06-26T10:54:22  [4315792768] INFO  DnsMadeEasyProvider[dnsmadeeasy] populate:   found 0 records, exists=True
2023-06-26T10:54:22  [4315792768] WARNING DnsMadeEasyProvider[dnsmadeeasy] root NS record supported, but no record is configured for yzguy.lll.
2023-06-26T10:54:22  [4315792768] INFO  DnsMadeEasyProvider[dnsmadeeasy] plan:   Creates=2, Updates=0, Deletes=0, Existing Records=0
2023-06-26T10:54:22  [4315792768] INFO  Plan
********************************************************************************
* yzguy.lll.
********************************************************************************
* dnsmadeeasy (DnsMadeEasyProvider)
*   Create <ARecord A 3600, record2.yzguy.lll., ['192.168.225.11']> (config)
*   Create <ARecord A 3600, record3.yzguy.lll., ['192.168.225.12']> (config)
*   Summary: Creates=2, Updates=0, Deletes=0, Existing Records=0
********************************************************************************
```

mintopia commented:
Thanks for reporting, can confirm and reproduce it here as well. Suspect it's the API not matching documentation for bulk deletions. I'll investigate further and update the PR when done.

mintopia commented:
Fixed the issue, can confirm it now only deletes the specified records.

```
(env) mintopia@DC-PW03TMGZ:/mnt/c/Projects/DNSMadeEasy/octodns$ octodns-sync --config-file=./config/test.yaml --doit
2023-06-26T16:05:49  [139984489189824] INFO  Manager __init__: config_file=./config/test.yaml (octoDNS 0.9.21)
2023-06-26T16:05:49  [139984489189824] INFO  Manager _config_executor: max_workers=1
2023-06-26T16:05:49  [139984489189824] INFO  Manager _config_include_meta: include_meta=False
2023-06-26T16:05:49  [139984489189824] INFO  Manager __init__: global_processors=[]
2023-06-26T16:05:49  [139984489189824] INFO  Manager __init__: provider=config (octodns.provider.yaml 0.9.21)
2023-06-26T16:05:49  [139984489189824] INFO  Manager __init__: provider=dnsmadeeasy (octodns_dnsmadeeasy 0.0.2+d83888b7)
2023-06-26T16:05:49  [139984489189824] INFO  Manager sync: eligible_zones=[], eligible_targets=[], dry_run=False, force=False, plan_output_fh=<stdout>
2023-06-26T16:05:49  [139984489189824] INFO  Manager sync:   zone=test1.dme.js42.io.
2023-06-26T16:05:49  [139984489189824] INFO  Manager sync:   sources=['config'] -> targets=['dnsmadeeasy']
2023-06-26T16:05:49  [139984489189824] INFO  YamlProvider[config] populate:   found 2 records, exists=False
2023-06-26T16:05:49  [139984489189824] INFO  DnsMadeEasyProvider[dnsmadeeasy] plan: desired=test1.dme.js42.io.
2023-06-26T16:05:50  [139984489189824] INFO  DnsMadeEasyProvider[dnsmadeeasy] populate:   found 3 records, exists=True
2023-06-26T16:05:50  [139984489189824] WARNING DnsMadeEasyProvider[dnsmadeeasy] root NS record supported, but no record is configured for test1.dme.js42.io.
2023-06-26T16:05:50  [139984489189824] INFO  DnsMadeEasyProvider[dnsmadeeasy] plan:   Creates=0, Updates=0, Deletes=1, Existing Records=3
2023-06-26T16:05:50  [139984489189824] INFO  Plan
********************************************************************************
* test1.dme.js42.io.
********************************************************************************
* dnsmadeeasy (DnsMadeEasyProvider)
*   Delete <CnameRecord CNAME 60, a.test1.dme.js42.io., example.com.>
*   Summary: Creates=0, Updates=0, Deletes=1, Existing Records=3
********************************************************************************


2023-06-26T16:05:50  [139984489189824] INFO  DnsMadeEasyProvider[dnsmadeeasy] apply: making 1 changes to test1.dme.js42.io.
2023-06-26T16:05:51  [139984489189824] INFO  Manager sync:   1 total changes
(env) mintopia@DC-PW03TMGZ:/mnt/c/Projects/DNSMadeEasy/octodns$ octodns-sync --config-file=./config/test.yaml --doit
2023-06-26T16:05:53  [139638264574400] INFO  Manager __init__: config_file=./config/test.yaml (octoDNS 0.9.21)
2023-06-26T16:05:53  [139638264574400] INFO  Manager _config_executor: max_workers=1
2023-06-26T16:05:53  [139638264574400] INFO  Manager _config_include_meta: include_meta=False
2023-06-26T16:05:53  [139638264574400] INFO  Manager __init__: global_processors=[]
2023-06-26T16:05:53  [139638264574400] INFO  Manager __init__: provider=config (octodns.provider.yaml 0.9.21)
2023-06-26T16:05:53  [139638264574400] INFO  Manager __init__: provider=dnsmadeeasy (octodns_dnsmadeeasy 0.0.2+d83888b7)
2023-06-26T16:05:53  [139638264574400] INFO  Manager sync: eligible_zones=[], eligible_targets=[], dry_run=False, force=False, plan_output_fh=<stdout>
2023-06-26T16:05:53  [139638264574400] INFO  Manager sync:   zone=test1.dme.js42.io.
2023-06-26T16:05:53  [139638264574400] INFO  Manager sync:   sources=['config'] -> targets=['dnsmadeeasy']
2023-06-26T16:05:53  [139638264574400] INFO  YamlProvider[config] populate:   found 2 records, exists=False
2023-06-26T16:05:53  [139638264574400] INFO  DnsMadeEasyProvider[dnsmadeeasy] plan: desired=test1.dme.js42.io.
2023-06-26T16:05:54  [139638264574400] INFO  DnsMadeEasyProvider[dnsmadeeasy] populate:   found 2 records, exists=True
2023-06-26T16:05:54  [139638264574400] WARNING DnsMadeEasyProvider[dnsmadeeasy] root NS record supported, but no record is configured for test1.dme.js42.io.
2023-06-26T16:05:54  [139638264574400] INFO  DnsMadeEasyProvider[dnsmadeeasy] plan:   No changes
2023-06-26T16:05:54  [139638264574400] INFO  Plan
********************************************************************************
No changes were planned
********************************************************************************
```

yzguy commented on tests/test_octodns_provider_dnsmadeeasy.py:
```diff
@@ -131,13 +131,18 @@ def test_apply(self):
             domains,
         ]
         plan = provider.plan(self.expected)
+        print(plan)
```
Seems like a leftover debugging line

yzguy commented:
@mintopia Excellent, just pulled your changes and confirmed it's working as expected.

There was an additional change/cleanup thing from #19 that if you're up to it could roll into this PR? 

* The `zone_records` method has a [try/except](https://github.com/octodns/octodns-dnsmadeeasy/pull/19/files#diff-944cb59ec539eafc04ab0fabfa9c499f5de77cd7cf1e4d88f60030d9d62cc58fL250-L253) that shouldn't be needed any more, that was basically catching when `self.domains.get` was called in the `client.records` returning False, then it attempted to make a call to `/False/records`, generating a `DnsMadeEasyClientNotFound` exception.

This should no longer happen because your change would add the newly created zone to the cache

mintopia commented:
The API deleting everything on a DELETE to /records with no ID specified - yes, scary, but semantically is correct - but I personally don't like having delete endpoints like that. At least with octodns you can easily create the records again!

I will raise the issue and see if we can get something put in place to reduce the risk of someone accidentally deleting their records.

I've incorporated the changes from PR #19 to remove the unneeded try/except. As part of updating the tests I've split the specific test (populating records from a non-existent zone) into its own test case as requests-mock was returning the first non-existent domain response for subsequent calls within the same test case.

yzguy commented:
@mintopia Thanks for the fix/features! Will cut a release here in a few
